Document Getting started procedure for Argo CD Rollouts via OpenShift GitOps (#78060)
openshift/openshift-docs: modules/gitops-manually-aborting-the-rollout.adoc

// Module included in the following assemblies:
//
// * argo_rollouts/getting-started-with-argo-rollouts.adoc

:_mod-docs-content-type: PROCEDURE
[id="gitops-manually-aborting-the-rollout_{context}"]
= Manually aborting the rollout

When using a canary deployment, the rollout deploys an initial canary version of the application. You can verify it either manually or programmatically. After you verify the canary version and promote it to stable, the new stable version is made available to all users.

> However, sometimes bugs, errors, or deployment issues are discovered in the canary version, and you might wish to abort the canary rollout and rollback to a stable version of your application.

Reviewer: ```suggestion
However, sometimes bugs, errors, or deployment issues are discovered in the canary version, and you might want to abort the canary rollout and rollback to a stable version of your application.
```
"want" is better; I am not sure how "wish" translates

Aborting a canary rollout deletes the resources of the new canary version and restores the previous stable version of your application. All network traffic such as ingress, route, or virtual service that was being directed to the canary returns to the original stable version.

The following example procedure deploys a new `red` canary version of your application, and then aborts it before it is fully promoted to stable.

.Procedure 

. Update the container image version and and modify the `.spec.template.spec.containers.image` value from `argoproj/rollouts-demo:yellow` to `argoproj/rollouts-demo:red` by running the following command in the Argo Rollouts CLI:
+
[source,terminal]
----
$ oc argo rollouts set image rollouts-demo rollouts-demo=argoproj/rollouts-demo:red -n <namespace> <1>
----
<1> Specify the namespace where the `Rollout` custom resource (CR) is defined.
+

> .Sample output

Reviewer: ```suggestion
.Example output
```
See https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output 
The guidance doesn't appear to bar "Sample" specifically, but the rest of the repo uses "Example"; so I would guess that someone would search for "Example", not "Sample"

[source,terminal]
----
rollout "rollouts-demo" image updated
----
+
The container image deployed in the rollout is modified and the rollout initiates a new canary deployment.

. Wait for the rollout to reach the paused status.

. Verify that the rollout deploys the `rollouts-demo:red` canary version and reaches the paused status by running the following command: 
+
[source,terminal]
----
$ oc argo rollouts get rollout rollouts-demo --watch -n <namespace> <1>
----
<1> Specify the namespace where the `Rollout` CR is defined.
+

> .Sample output

Reviewer: ```suggestion
.Example output
```

├──# revision:3                                                             
│  └──⧉ rollouts-demo-5747959bdb           ReplicaSet  ✔ Healthy     75s    canary
│     └──□ rollouts-demo-5747959bdb-fdrsg  Pod         ✔ Running     75s    ready:1/1
├──# revision:2                                                             
│  └──⧉ rollouts-demo-6cf78c66c5           ReplicaSet  ✔ Healthy     9m45s  stable
│     ├──□ rollouts-demo-6cf78c66c5-zrgd4  Pod         ✔ Running     9m44s  ready:1/1
│     ├──□ rollouts-demo-6cf78c66c5-2ptpp  Pod         ✔ Running     4m58s  ready:1/1
│     ├──□ rollouts-demo-6cf78c66c5-tmk6c  Pod         ✔ Running     4m38s  ready:1/1
│     └──□ rollouts-demo-6cf78c66c5-zv6lx  Pod         ✔ Running     4m27s  ready:1/1
└──# revision:1                                                             
   └──⧉ rollouts-demo-687d76d795           ReplicaSet  • ScaledDown  17m   
----

. Abort the update of the rollout by running the following command:
+
[source,terminal]
----
$ oc argo rollouts abort rollouts-demo -n <namespace> <1>
----
<1> Specify the namespace where the `Rollout` CR is defined.
+

> .Sample output

Reviewer: ```suggestion
.Example output
```

[source,terminal]
----
rollout 'rollouts-demo' aborted
----
+
The Argo Rollouts controller deletes the canary resources of the application, and rolls back to the stable version. 

. Verify that after aborting the rollout, now the canary `ReplicaSet` is scaled to 0 replicas by running the following command: 
+
[source,terminal]
----
$ oc argo rollouts get rollout rollouts-demo --watch -n <namespace> <1>
----
<1> Specify the namespace where the `Rollout` CR is defined.
+

> .Sample output

Reviewer: ```suggestion
.Example output
```

│     ├──□ rollouts-demo-6cf78c66c5-zv6lx  Pod         ✔ Running     10m    ready:1/1
│     └──□ rollouts-demo-6cf78c66c5-mlbsh  Pod         ✔ Running     4m47s  ready:1/1
└──# revision:1                                                             
   └──⧉ rollouts-demo-687d76d795           ReplicaSet  • ScaledDown  24m 
----
+
The rollout status is marked as `Degraded` indicating that even though the application has rolled back to the previous stable version, `yellow`, the rollout is not currently at the wanted version, `red`, that was set within the `.spec.template.spec.containers.image` field. 
+
[NOTE]
====
The `Degraded` status does not reflect the health of the application. It only indicates that there is a mismatch between the wanted and running container image versions. 
====

. Update the container image version to the previous stable version, `yellow`, and modify the `.spec.template.spec.containers.image` value by running the following command:
+
[source,terminal]
----
$ oc argo rollouts set image rollouts-demo rollouts-demo=argoproj/rollouts-demo:yellow -n <namespace> <1>
----
<1> Specify the namespace where the `Rollout` CR is defined.
+

> .Sample output

Reviewer: same

[source,terminal]
----
rollout "rollouts-demo" image updated
----
+

> The rollout skips the analysis and promotion steps, rollsback to the previous stable version, `yellow`, and fast-tracks the deployment of the stable `ReplicaSet`.

Reviewer: ```suggestion
The rollout skips the analysis and promotion steps, rolls back to the previous stable version, `yellow`, and fast-tracks the deployment of the stable `ReplicaSet`.
```

. Verify that the rollout status is immediately marked as `Healthy` by running the following command:
+
[source,terminal]
----
$ oc argo rollouts get rollout rollouts-demo --watch -n <namespace> <1>
----
<1> Specify the namespace where the `Rollout` CR is defined.
+

> .Sample output

Reviewer: same
